core/src/main/java/org/apache/gravitino/authorization/PermissionManager.java

```diff
@@ -373,11 +377,253 @@ User revokeRolesFromUser(String metalake, List<String> roles, String user) {
     }
   }
 
-  private List<String> toRoleNames(List<RoleEntity> roleEntities) {
-    return roleEntities.stream().map(RoleEntity::name).collect(Collectors.toList());
+  Role grantPrivilegesToRole(
+      String metalake, String role, MetadataObject object, List<Privilege> privileges) {
+    try {
+      AuthorizationPluginCallbackWrapper authorizationPluginCallbackWrapper =
+          new AuthorizationPluginCallbackWrapper();
+      Role updatedRole =
+          store.update(
+              AuthorizationUtils.ofRole(metalake, role),
+              RoleEntity.class,
+              Entity.EntityType.ROLE,
+              roleEntity -> {
+                List<SecurableObject> updateSecurableObjects =
+                    updateSecurableObjects(
+                        roleEntity.securableObjects(),
+                        object,
+                        oldObject -> {
+                          if (oldObject == null) {
+                            // Add a new securable object if there not exists the object in the role
+                            SecurableObject securableObject =
+                                SecurableObjects.parse(
+                                    object.fullName(),
+                                    object.type(),
+                                    Lists.newArrayList(privileges));
+
+                            authorizationPluginCallbackWrapper.setCallBack(
+                                () ->
+                                    AuthorizationUtils.callAuthorizationPluginForMetadataObject(
+                                        metalake,
+                                        object,
+                                        authorizationPlugin -> {
+                                          authorizationPlugin.onRoleUpdated(
+                                              roleEntity,
+                                              RoleChange.addSecurableObject(role, securableObject));
+                                        }));
+
+                            return securableObject;
+                          } else {
+                            // Removed duplicated privileges by set
+                            Set<Privilege> updatePrivileges = Sets.newHashSet();
+                            updatePrivileges.addAll(oldObject.privileges());
+                            // If old object contains all the privileges to grant, the object don't
+                            // need to change.
+                            if (updatePrivileges.containsAll(privileges)) {
+                              return oldObject;
+                            } else {
+                              updatePrivileges.addAll(privileges);
```

Comment:
Should you check the `allow` and `deny` of the privileges here?

Also, do you need to check if the privileges are inclusive or exclusive, so you don't have store some particular privileges to the storage?

Reply:
1. If the privilege have the same Name , but it have different conditions. They are different privileges.
2. I don't know the meaning about `inclusive or exclusive`.
If you mean the case `use schema` and `select table`, they don't check it in the granting or revoking operations.

Reply:
OK, I will check the privileges. We don't allow the same name privilege is allowed and denied at the same time.